src/components/ImageViewerContainer.ts

```diff
@@ -0,0 +1,138 @@
+import { Component, DOM, createElement } from "react";
+
+import { Alert } from "./Alert";
+import { ImageViewer } from "./ImageViewer";
+import { UrlHelper } from "../UrlHelper";
+
+interface WrapperProps {
+    class: string;
+    mxObject?: mendix.lib.MxObject;
+    style: string;
+    readOnly: boolean;
+}
+
+interface ImageViewerContainerProps extends WrapperProps {
+    source: DataSource;
+    imageAttribute: string;
+    urlStatic: string;
+    imageStatic: string;
+    width: number;
+    widthUnit: Units;
+    height: number;
+    heightUnit: Units;
+}
+
+interface ImageViewerContainerState {
+    alertMessage?: string;
+    imageUrl: string;
+    isLoading?: boolean;
+}
+
+type DataSource = "systemImage" | "urlAttribute" | "staticUrl" | "staticImage";
+type Units = "auto" | "pixels" | "percentage";
+
+class ImageViewerContainer extends Component<ImageViewerContainerProps, ImageViewerContainerState> {
+    private subscriptionHandle: number;
+    private attributeCallback: (mxObject: mendix.lib.MxObject) => () => void;
+
+    constructor(props: ImageViewerContainerProps) {
+        super(props);
+
+        const alertMessage = ImageViewerContainer.validateProps(props);
+        this.state = {
+            alertMessage,
+            imageUrl: "",
+            isLoading: true
+        };
+        this.attributeCallback = mxObject => () => this.getImageUrl(mxObject);
+    }
+
+    render() {
+        if (this.state.alertMessage) {
+            return createElement(Alert, { message: this.state.alertMessage });
+        }
+        if (this.state.isLoading) {
+            return DOM.div(null,
+                DOM.i({ className: "glyphicon glyphicon-cog glyph-spin" }),
+                DOM.span(null, " Loading ...")
+            );
+        }
+
+        return createElement(ImageViewer, {
+            height: this.props.height,
+            heightUnits: this.props.heightUnit,
+            imageurl: this.state.imageUrl,
+            width: this.props.width,
+            widthUnits: this.props.widthUnit
```

Comment:
Can destructure all these and reduce this to two lines...
```
const { height, heightUnit, width, widthUnit } = this.props;
return createElement(ImageViewer, { height, heightUnits, imageurl: this.state.imageUrl, width, widthUnits });
```

Destructure the state too to make it cleaner, and of course synchronize your naming... Why have heightUnit in one place and heightUnits in another? It's units.